rclcpp/test/benchmark/benchmark_executor.cpp

```diff
@@ -0,0 +1,353 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "performance_test_fixture/performance_test_fixture.hpp"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/scope_exit.hpp"
+#include "test_msgs/msg/empty.hpp"
+
+using namespace std::chrono_literals;
+using performance_test_fixture::PerformanceTest;
+
+class PerformanceTestExecutor : public PerformanceTest
+{
+public:
+  void SetUp(benchmark::State & st)
+  {
+    rclcpp::init(0, nullptr);
+    callback_count = 0;
+    node = std::make_shared<rclcpp::Node>("my_node");
+
+    publisher = node->create_publisher<test_msgs::msg::Empty>(
+      "empty_msgs", rclcpp::QoS(10));
+    auto callback = [this](test_msgs::msg::Empty::SharedPtr) {this->callback_count++;};
+    subscription =
+      node->create_subscription<test_msgs::msg::Empty>(
+      "empty_msgs", rclcpp::QoS(10), std::move(callback));
+
+    PerformanceTest::SetUp(st);
+  }
+  void TearDown(benchmark::State & st)
+  {
+    PerformanceTest::TearDown(st);
+    publisher.reset();
+    subscription.reset();
+    node.reset();
+    rclcpp::shutdown();
+  }
+
+  test_msgs::msg::Empty empty_msgs;
+  rclcpp::Node::SharedPtr node;
+  rclcpp::Publisher<test_msgs::msg::Empty>::SharedPtr publisher;
+  rclcpp::Subscription<test_msgs::msg::Empty>::SharedPtr subscription;
+  int callback_count;
+};
+
+BENCHMARK_F(PerformanceTestExecutor, single_thread_executor)(benchmark::State & st)
+{
+  rclcpp::executors::SingleThreadedExecutor executor;
+  executor.add_node(node);
+
+  reset_heap_counters();
+
+  for (auto _ : st) {
+    st.PauseTiming();
+    publisher->publish(empty_msgs);
+    st.ResumeTiming();
+
+    executor.spin_some(100ms);
+
+    benchmark::ClobberMemory();
+  }
+  if (callback_count == 0) {
+    st.SkipWithError("No message was received");
+  }
+}
+
+BENCHMARK_F(PerformanceTestExecutor, multi_thread_executor)(benchmark::State & st)
+{
+  rclcpp::executors::MultiThreadedExecutor executor;
+  executor.add_node(node);
+
+  reset_heap_counters();
+
+  for (auto _ : st) {
+    st.PauseTiming();
+    publisher->publish(empty_msgs);
+    st.ResumeTiming();
+
+    executor.spin_some(100ms);
+
+    benchmark::ClobberMemory();
```

Comment:
Does this have any effect without `benchmark::DoNotOptimize()` as well? @brawner?

Reply:
Here's the explanation from benchmark's readme:

> The second tool for preventing optimizations is ClobberMemory(). In essence ClobberMemory() forces the compiler to perform all pending writes to global memory. Memory managed by block scope objects must be "escaped" using DoNotOptimize(...) before it can be clobbered. In the below example ClobberMemory() prevents the call to v.push_back(42) from being optimized away.

It sounds like clobbermemory has no effect if there is not a DoNotOptimize